pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,635 @@
+			clientset := fakeclientset.NewSimpleClientset()
+			if tt.existingAgentCRD != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingAgentCRD)
+			}
+			tt.prepareReactor(clientset)
+			controller := newControllerMonitor(clientset)
+			initController(controller)
+			err := controller.controllerMonitor.createAgentCRD(tt.name)
+			if tt.expectedError != "" {
+				assert.Equal(t, tt.expectedError, err.Error())
+			} else {
+				assert.Nil(t, err)
+				crd, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().Get(context.TODO(), tt.expectedAgentCRD.Name, metav1.GetOptions{})
+				assert.NoError(t, err)
+				assert.Equal(t, tt.expectedAgentCRD, crd)
+			}
+		})
+	}
+}
+
+func TestDeleteStaleAgentCRD(t *testing.T) {
+	tc := []struct {
+		name                 string
+		agentCRDs            []string
+		existingExternalNode *v1alpha1.ExternalNode
+		prepareReactor       func(clientset *fakeclientset.Clientset)
+		expected             []string
+	}{
+		{
+			name:           "No external Nodes present",
+			agentCRDs:      []string{"TestAgentCRD1", "TestAgentCRD2", "TestAgentCRD3", "TestAgentCRD4"},
+			prepareReactor: func(clientset *fakeclientset.Clientset) {},
+			expected:       []string{},
+		},
+		{
+			name: "No Agent CRDS present",
+			prepareReactor: func(clientset *fakeclientset.Clientset) {
+				clientset.PrependReactor("list", "antreaagentinfos", func(action cgtesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, &v1beta1.AntreaAgentInfoList{}, errors.New("error getting agent crds")
+				})
+			},
+			expected: []string{},
+		},
+		{
+			name:                 "Both Agent CRDs and external Node present",
+			agentCRDs:            []string{"TestAgentCRD1", "TestAgentCRD2"},
+			existingExternalNode: &v1alpha1.ExternalNode{ObjectMeta: metav1.ObjectMeta{Name: "TestAgentCRD1", Namespace: "ns1", Labels: map[string]string{"en": "vm2"}}},
+			prepareReactor:       func(clientset *fakeclientset.Clientset) {},
+			expected:             []string{"TestAgentCRD1"},
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			clientset := fakeclientset.NewSimpleClientset()
+			if tt.existingExternalNode != nil {
+				clientset = fakeclientset.NewSimpleClientset(tt.existingExternalNode)
+			}
+			tt.prepareReactor(clientset)
+			controller := newControllerMonitor(clientset)
+			initController(controller)
+			controller.controllerMonitor.externalNodeEnabled = true
+			for _, crd := range tt.agentCRDs {
+				agentCRD := new(v1beta1.AntreaAgentInfo)
+				agentCRD.Name = crd
+				agentCRD.ResourceVersion = "0"
+				_, err := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().Create(context.TODO(), agentCRD, metav1.CreateOptions{})
+				assert.NoError(t, err)
+			}
+			controller.controllerMonitor.deleteStaleAgentCRDs()
+			crds, _ := controller.controllerMonitor.client.CrdV1beta1().AntreaAgentInfos().List(context.TODO(), metav1.ListOptions{ResourceVersion: "0"})
```

Comment:
Better to assign the returned error to a variable and make sure it's nil. You may get an error when you didn't prepare the test data properly, then you can catch it here.